command/ssh/certificate.go

```diff
@@ -429,3 +458,27 @@ type oidcPayload struct {
 	jose.Claims
 	Email string `json:"email"` // OIDC indicator
 }
+
+func deriveMachineID() (uuid.UUID, error) {
+	// use /etc/machine-id
+	machineID, err := ioutil.ReadFile("/etc/machine-id")
+	if err != nil {
+		return uuid.Nil, err
+	}
+
+	// 32 bytes, not secret
+	key := []byte("moon machines mortify more minds")
```

Comment:
The block size of SHA256 is 64 bytes, that is the recommended size of the key.

Reply:
What do you think about this: https://github.com/smallstep/cli/pull/230?

Reply:
We should not use SHA1
https://latacora.micro.blog/2018/04/03/cryptographic-right-answers.html

Reply:
We are not trying to prevent collision attacks here. We just need to derive an id based off the machine-id plus a fixed "namespace" we specify that has a high probability of of being unique.

Reply:
Well, the other implementation was closer to what `systemd` methods do.
https://github.com/systemd/systemd/blob/master/src/libsystemd/sd-id128/sd-id128.c#L273

Reply:
Okay sure.

Reply:
@maraino I've looked into this. systemd uses the `machine-id` as the key and an application ID as the message. That is, a `16 byte` key, half of what's proposed here. Further, none of this matters because we're truncating the resulting hash so an attack attempting to recover the `machine-id` is not really possible. What we loose by truncating the sha is entropy space for our resulting uuid.

Anyway, sha3 and blake2 are designed to be used trivially as MACs, so I'm just going to use blake2: https://godoc.org/golang.org/x/crypto/blake2b#New I'm going to use the first 16 bytes of the machine-id as the key and a 16byte fixed value as salt so we get a hash over 32 bytes of data. Here's some more color: https://crypto.stackexchange.com/questions/1186/is-hklengthx-a-secure-mac-construction

Reply:
https://github.com/smallstep/cli/pull/230/commits/c15a33968749a7abce1ef1879d69b8128d9d99b5